efi/fw_load_handler.go

```diff
@@ -236,22 +237,90 @@ func (h *fwLoadHandler) measureBootManagerCodePreOS(ctx pcrBranchContext) error
 	// OS-present environment as a separator event measured to PCRs 0-7, but EDK2 measures a separator to
 	// PCR7 as soon as the secure boot policy is measured and system preparation applications are considered
 	// part of the pre-OS environment - they are measured to PCR4 before the pre-OS to OS-present transition
-	// is signalled by measuring separators to the remaining PCRs. This seems sensible, but newer Dell
-	// devices load an agent from firmware before shim is executed and measure this to PCR4 as part of the
-	// OS-present environment, which seems wrong. The approach here assumes that the EDK2 behaviour is
-	// correct.
-	for _, event := range h.log.Events {
+	// is signalled by measuring separators to the remaining PCRs. The UEFI specification says that system
+	// preparation applications are executed before the ready to boot signal, which is when the transition
+	// from pre-OS to OS-present occurs, so I think we can be confident that we're correct here.
+	events := h.log.Events
+	measuredSeparator := false
+	for len(events) > 0 {
+		event := events[0]
+		events = events[1:]
+
 		if event.PCRIndex != tcglog.PCRIndex(bootManagerCodePCR) {
 			continue
 		}
 
 		if event.EventType == tcglog.EventTypeSeparator {
-			return h.measureSeparator(ctx, bootManagerCodePCR, event)
+			if err := h.measureSeparator(ctx, bootManagerCodePCR, event); err != nil {
+				return err
+			}
+			measuredSeparator = true
+			break
 		}
 		ctx.ExtendPCR(bootManagerCodePCR, tpm2.Digest(event.Digests[ctx.PCRAlg()]))
 	}
 
-	return errors.New("missing separator")
+	if !measuredSeparator {
+		return errors.New("missing separator")
+	}
+
+	// Some newer laptops including those from Dell and Lenovo execute code from a firmware volume as part
+	// of the OS-present environment, before shim runs, and using the LoadImage API which results in an
+	// additional measurement to PCR4. Copy this into the profile if it's part of a well-known endpoint
+	// management application, else return an error. Anything else here will be picked up by the pre-install
+	// checks.
```

Comment:
I suppose it doesn't make sense to make this behavior of supporting these payload a policy option?

Reply:
I'd rather not make this an option for now - as far as we can tell, Absolute seems to be the only problematic firmware component we know about so far and the checks that will be added later will fail if anything else ends up being measured to PCR4 before the OS. I think we're probably better off adding additional accommodations for other components as and when we encounter them, where appropriate.

Reply:
ok